File: backend/app/models/audit.py
from datetime import UTC, datetime
from enum import StrEnum

from sqlalchemy import DDL, CheckConstraint, Index, event
from sqlmodel import Field, SQLModel

from app.core.types import UtcDateTime

# D-007 §8 — the complete audit action catalog. One row per semantic action.
AUDIT_ACTIONS = (
    "LOGIN_SUCCESS",
    "LOGIN_FAILURE",
    "LOGOUT",
    "ALERT_CONFIRM",
    "ALERT_DISMISS",
    "ALERT_RESOLVE",
    "ALERT_CORRECTION",
    "ALERT_SNOOZE",
    "CAMERA_CREATE",
    "CAMERA_UPDATE",
    "CAMERA_ENABLE",
    "CAMERA_DISABLE",
    "CAMERA_DELETE",
    "CAMERA_RESTORE",
    "REPORT_EXPORT",
    "AUDIT_EXPORT",
    "USER_CREATE",
    "USER_UPDATE",
    "USER_ENABLE",
    "USER_DISABLE",
    "USER_ROLE_CHANGE",
    "USER_PASSWORD_RESET",
    "USER_PROFILE_UPDATE",
    "USER_PASSWORD_CHANGE",
    "ALARM_SETTINGS_UPDATE",
    "BACKUP_TRIGGER",
    "RESTORE_TRIGGER",
)

_AUDIT_ACTIONS_SQL = ", ".join(f"'{action}'" for action in AUDIT_ACTIONS)

# A proper enum type over the same catalog, generated from the single
# AUDIT_ACTIONS source of truth so the two can never drift. Used for typed
# query-parameter validation on the audit viewer (P2 Step 11) — an unknown
# action value becomes a normal 422 instead of a silently-empty filter.
AuditAction = StrEnum("AuditAction", {action: action for action in AUDIT_ACTIONS})


class AuditLog(SQLModel, table=True):
    """D-007 — append-only audit trail. No pruning, ever."""

    __tablename__ = "audit_log"
    __table_args__ = (
        CheckConstraint(
            "actor_type IN ('user', 'system')", name="ck_audit_actor_type_valid"
        ),
        CheckConstraint(
            f"action IN ({_AUDIT_ACTIONS_SQL})", name="ck_audit_action_valid"
        ),
        CheckConstraint(
            "result IN ('success', 'denied', 'failure')", name="ck_audit_result_valid"
        ),
        Index("ix_audit_created_at", "created_at"),
        Index("ix_audit_action_created_at", "action", "created_at"),
        Index("ix_audit_user_created_at", "user_id", "created_at"),
        Index("ix_audit_target", "target_type", "target_ref"),
    )

    audit_id: int | None = Field(default=None, primary_key=True)
    actor_type: str
    user_id: int | None = Field(
        default=None, foreign_key="user.user_id", ondelete="RESTRICT"
    )
    username: str | None = Field(default=None)
    role: str | None = Field(default=None)
    action: str
    target_type: str | None = Field(default=None)
    target_ref: str | None = Field(default=None)
    result: str
    detail: str | None = Field(default=None)
    request_id: str | None = Field(default=None)
    source_ip: str | None = Field(default=None)
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC), sa_type=UtcDateTime
    )


# Immutability triggers (D-007, NFR-21). SQLModel/SQLAlchemy cannot express
# these declaratively, so they're attached as raw DDL fired right after the
# table itself is created — under create_all() today and under Alembic once
# P9 lands.
_NO_UPDATE_TRIGGER = DDL(
    """
    CREATE TRIGGER trg_audit_log_no_update BEFORE UPDATE ON audit_log
    BEGIN SELECT RAISE(ABORT, 'audit_log is append-only'); END;
    """
)

_NO_DELETE_TRIGGER = DDL(
    """
    CREATE TRIGGER trg_audit_log_no_delete BEFORE DELETE ON audit_log
    BEGIN SELECT RAISE(ABORT, 'audit_log is append-only'); END;
    """
)

# Public so app/dev/wipe.py can drop and *re-execute these same objects*
# rather than re-typing the SQL. NFR-21's append-only guarantee must have
# exactly one source of truth: a second copy of this DDL that drifted from
# this one would be undetectable until it mattered.
AUDIT_IMMUTABILITY_TRIGGERS = (_NO_UPDATE_TRIGGER, _NO_DELETE_TRIGGER)
AUDIT_IMMUTABILITY_TRIGGER_NAMES = (
    "trg_audit_log_no_update",
    "trg_audit_log_no_delete",
)

for _trigger in AUDIT_IMMUTABILITY_TRIGGERS:
    event.listen(AuditLog.__table__, "after_create", _trigger)

File: frontend/src/api/audit.ts
import api from "@/api/client"
import { downloadBlobResponse } from "@/utils/download"
import type { ExportFormat } from "@/components/ui/ExportButton"

/**
 * The 27-entry action catalog (`AUDIT_ACTIONS`, `app/models/audit.py`),
 * mirrored here rather than fetched — there is no options endpoint for it
 * (only `GET /` and `GET /export` exist on this router), and unlike
 * Phase 14's alarm-sound allowlist this one is a fixed, migration-gated
 * StrEnum backed by a database CHECK constraint, not runtime config that
 * could silently drift. The plan's own PR body says as much: "the client
 * mirrors that catalog rather than accepting free text." An unlisted value
 * is still a 422 server-side regardless of what this list contains.
 */
export const AUDIT_ACTIONS = [
  "LOGIN_SUCCESS",
  "LOGIN_FAILURE",
  "LOGOUT",
  "ALERT_CONFIRM",
  "ALERT_DISMISS",
  "ALERT_RESOLVE",
  "ALERT_CORRECTION",
  "ALERT_SNOOZE",
  "CAMERA_CREATE",
  "CAMERA_UPDATE",
  "CAMERA_ENABLE",
  "CAMERA_DISABLE",
  "CAMERA_DELETE",
  "CAMERA_RESTORE",
  "REPORT_EXPORT",
  "AUDIT_EXPORT",
  "USER_CREATE",
  "USER_UPDATE",
  "USER_ENABLE",
  "USER_DISABLE",
  "USER_ROLE_CHANGE",
  "USER_PASSWORD_RESET",
  "USER_PROFILE_UPDATE",
  "USER_PASSWORD_CHANGE",
  "ALARM_SETTINGS_UPDATE",
  "BACKUP_TRIGGER",
  "RESTORE_TRIGGER",
] as const

export type AuditAction = (typeof AUDIT_ACTIONS)[number]

/** `AuditResult` (`app/models/enums.py`) — a 3-value StrEnum, same reasoning as AUDIT_ACTIONS. */
export const AUDIT_RESULTS = ["success", "denied", "failure"] as const
export type AuditResult = (typeof AUDIT_RESULTS)[number]

/**
 * `target_type` has no enum on the backend — it's a free `str` column, one
 * literal string per call site (`audit.record(..., target_type="camera")`
 * and friends). Enumerated here by grepping every such call site rather
 * than guessed: `backup`, `camera`, `export`, `incident`, `restore`,
 * `session`, `user`. Same reasoning as AUDIT_ACTIONS — these are literal
 * strings in the backend's own source, not data that could add a new value
 * without a code change, so mirroring them is a real reflection of the
 * backend, not a stand-in for one.
 */
export const AUDIT_TARGET_TYPES = [
  "backup",
  "camera",
  "export",
  "incident",
  "restore",
  "session",
  "user",
] as const

/**
 * The six fields `AUDIT_SORT_FIELDS` (`routes/audit.py:28`) allows.
 * `audit_id` is always applied as a tie-break server-side even when sorting
 * by something else, so equal-`created_at` rows never shuffle between pages.
 * An unlisted value is a 422, same rule as Phase 7's `ALERT_SORT_FIELDS`.
 */
export const AUDIT_SORT_FIELDS = [
  "audit_id",
  "created_at",
  "action",
  "result",
  "user_id",
  "target_type",
] as const

export type AuditSortField = (typeof AUDIT_SORT_FIELDS)[number]
export type SortOrder = "asc" | "desc"

export interface AuditLogEntry {
  audit_id: number
  actor_type: string
  user_id: number | null
  username: string | null
  role: string | null
  action: string
  target_type: string | null
  target_ref: string | null
  result: string
  detail: Record<string, unknown> | null
  request_id: string | null
  source_ip: string | null
  created_at: string
}

export interface AuditLogListResponse {
  total_filtered: number
  items: AuditLogEntry[]
}

export interface GetAuditLogsParams {
  action?: AuditAction[]
  user_id?: number[]
  result?: AuditResult[]
  target_type?: string
  /**
   * Declared, intentionally unsent: `target_ref` is read and displayed
   * already (`AuditLog.tsx:409`), but nothing on the filter side sets it —
   * the existing `search` box already matches against the backend's
   * `target_ref`, so a dedicated filter field would just duplicate it.
   */
  target_ref?: string
  start_date?: string
  end_date?: string
  search?: string
  sort_by?: AuditSortField
  sort_order?: SortOrder
  limit?: number
  offset?: number
}

export async function getAuditLogs(params: GetAuditLogsParams) {
  const { data } = await api.get<AuditLogListResponse>("/audit-logs/", { params })
  return data
}

/**
 * `GET /api/audit-logs/export` takes the same filter set as the list route
 * plus `sort_by`/`sort_order`/`format`, and — confirmed directly against
 * `routes/audit.py:214` — genuinely accepts `format=csv|pdf` on this
 * synchronous route today, real PDF included, no async job required. This
 * export is itself audited (AUDIT_EXPORT), using a watermark so that row
 * can never appear inside the file it produces; viewing the log is not.
 */
export async function exportAuditLogs(params: GetAuditLogsParams, format: ExportFormat = "csv") {
  const response = await api.get<Blob>("/audit-logs/export", {
    params: { ...params, format },
    responseType: "blob",
  })

  downloadBlobResponse(response, `adas_audit_export.${format}`)
}
